beater/beater_test.go

```diff
@@ -0,0 +1,159 @@
+//       this can be removed.
+var testCount int
+
+type DummyOutputClient struct {
+}
+
+func (d *DummyOutputClient) Publish(batch pubs.Batch) error {
+	go batch.ACK()
+	return nil
+}
+
+func (d *DummyOutputClient) Close() error {
+	return nil
+}
+
+func SetupServer(b *testing.B) *http.ServeMux {
+	out := outputs.Group{
+		Clients:   []outputs.Client{&DummyOutputClient{}},
+		BatchSize: 5,
+		Retry:     0, // no retry. on error drop events
+	}
+
+	queueFactory := func(e queue.Eventer) (queue.Queue, error) {
+		return memqueue.NewBroker(memqueue.Settings{
+			Eventer: e,
+			Events:  20,
+		}), nil
+	}
+	testCount++
+	pip, err := pipeline.New(
+		monitoring.Default.NewRegistry("testing"+string(testCount)),
+		queueFactory, out, pipeline.Settings{
+			WaitClose:     0,
+			WaitCloseMode: pipeline.NoWaitOnClose,
+		})
+
+	if err != nil {
+		b.Fatalf("error initializing publisher: %v", err)
+	}
+
+	pub, err := newPublisher(pip, 1)
+
+	if err != nil {
+		b.Fatal(err)
+	}
+	return newMuxer(defaultConfig, pub.Send)
+}
+
+func pluralize(entity string) string {
+	return entity + "s"
+}
+func createPayload(entitytype string, numEntities int) []byte {
+	data, err := tests.LoadValidData(entitytype)
+	if err != nil {
+		panic(err)
+	}
+	var payload map[string]interface{}
+	err = json.Unmarshal(data, &payload)
+	var entityList []interface{}
+	testEntities := payload[pluralize(entitytype)].([]interface{})
+
+	for i := 0; i < numEntities; i++ {
+		entityList = append(entityList, testEntities[i%len(testEntities)])
+	}
+	payload[pluralize(entitytype)] = entityList
+	out, err := json.Marshal(payload)
+	if err != nil {
+		panic(err)
+	}
+	return out
+}
+
+func benchmarkVariableSizePayload(b *testing.B, entitytype string, entries int) {
+	url := "/v1/" + pluralize(entitytype)
+	mux := SetupServer(b)
+	data := createPayload(entitytype, entries)
+	b.Logf("Using payload size: %d", len(data))
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		req, err := http.NewRequest("POST", url, bytes.NewReader(data))
+		req.Header.Add("Content-Type", "application/json")
+		if err != nil {
+			b.Error(err)
+		}
+
+		w := httptest.NewRecorder()
+		mux.ServeHTTP(w, req)
+
+		if w.Code != 202 {
+			b.Fatal(w.Body.String())
+		}
+	}
+}
+
+func BenchmarkTransactionsEndToEnd100Entries(b *testing.B) {
+	benchmarkVariableSizePayload(b, "transaction", 100)
+}
+
+func BenchmarkErrorsEndToEnd100Entries(b *testing.B) {
+	benchmarkVariableSizePayload(b, "error", 100)
+}
+
+func BenchmarkTransactionsEndToEnd1000Entries(b *testing.B) {
+	benchmarkVariableSizePayload(b, "transaction", 1000)
+}
+
+func BenchmarkErrorsEndToEnd1000Entries(b *testing.B) {
+	benchmarkVariableSizePayload(b, "error", 1000)
+}
+
+func BenchmarkTransactionsEndToEnd10000Entries(b *testing.B) {
+	benchmarkVariableSizePayload(b, "transaction", 10000)
+}
+
+func BenchmarkErrorsEndToEnd10000Entries(b *testing.B) {
+	benchmarkVariableSizePayload(b, "error", 10000)
+}
```

Comment:
these can be put into a table like:

```
func BenchmarkServer(b *testing.B) {
    entitieTypes := []string{"transaction", "error"}
    sizes := []int{100, 1000, 10000}

    for _, et := range entityTypes {
      b.run(et, func(b *testing.B) {
        for _, sz := range sizes {
          b.run(fmt.Sprintf("size=%v", sz), func(b *testing.B) {
            benchmarkVariableSizePayload(b, et, sz)
          })
        }
      })
    }
}
```

the names passed to `Run` are separated via `/`. One can filter for particular tests of 1000 events only by using `-bench "//1000"`